[CORE-12458] Prevent segment appender from writing into already dispatched head buffer

mmaslankaprv opened the pull request:
This commit introduces a logic that prevents the head chunk from being
appended if it is dispatched i.e. its content is being DMA written.

Previously the head chunk buffer might have been passed to the
`file::dma_write_call` but appends could still write to that chunk. This
caused a problem which manifested on some filesystems and some hardware
RAID controllers. The problem manifested as corruption in the last page
of a file on disk.

This PR prevents the head being written to disk from being concurrently
appended. The mechanism used here is copying the reminder which wasn't
flushed and spans beyond the last page aligned address to a new chunk.
The reminder is copied only if the current head write was dispatched.
After copying the chunk writer is requested to recycle the chunk and
return it to the cache.
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
### Improvements
- potentially better behavior when working with some hardware RAID controllers and portworx

mmaslankaprv commented:
/dt

mmaslankaprv commented:
/dt

mmaslankaprv committed:
s/chunk: add `copy_reminder_from` method to segment appender

Added method that copies the reminder that spans from the last page
aligned address to the end of the buffer. The method copies data and
sets the chunk positions to make sure the number of not flushed bytes is
preserved.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
s/appender: do not append to chunk which is being dma_written

This commit introduces a logic that prevents the head chunk from being
appended if it is dispatched i.e. its content is being DMA written.

Previously the head chunk buffer might have been passed to the
`file::dma_write_call` but appends could still write to that chunk. This
caused a problem which manifested on some filesystems and some hardware
RAID controllers. The problem manifested as corruption in the last page
of a file on disk.

This PR prevents the head being written to disk from being concurrently
appended. The mechanism used here is copying the reminder which wasn't
flushed and spans beyond the last page aligned address to a new chunk.
The reminder is copied only if the current head write was dispatched.
After copying the chunk writer is requested to recycle the chunk and
return it to the cache.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
s/tests: added test verifying reminder copy

Signed-off-by: Michał Maślanka <michal@redpanda.com>

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/75218-->
<!--71625a6e-e6a3-4a5d-b552-199fc4b6d1bf-->
<!--title-->
## Retry command for [Build#75218](https://buildkite.com/redpanda/redpanda/builds/75218)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/random_node_operations_smoke_test.py::RedpandaNodeOperationsSmokeTest.test_node_ops_smoke_test@{"cloud_storage_type":1,"mixed_versions":true}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--2732a1e1-9566-40b4-ba57-e0500ad57a2f-->
<!--TEST RESULTS-->
<!--0fb91bd1-04af-4526-a97d-c1defbc35356-->
<!--TEST RESULTS-->
<!--4b0f7681-efda-41a9-a945-b11bb6ea7da3-->
<!--TEST RESULTS-->
<!--cb74d1d2-d605-42ab-be3d-48e10ea178ee-->
<!--TEST RESULTS-->
<!--2921cd18-a714-49f0-b073-7f17999ca054-->
<!--TEST RESULTS-->
<!--71ccf048-fc0f-4d4e-9ca4-0e6e5bac86d4-->
<!--TEST RESULTS-->
<!--da3f7369-b534-4d0f-89a1-24fc218173db-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75218">build#75218</a></summary>


| test_class                                                                 | test_method                                                                | test_arguments                                     | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                           |
|:---------------------------------------------------------------------------|:---------------------------------------------------------------------------|:---------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| log_segment_appender_test                                                  | test_concurrent_append_flush                                               |                                                    | unit        | https://buildkite.com/redpanda/redpanda/builds/75218#019a309f-5e49-458f-9734-29dc7d1c5fba | FAIL          | 0/1      |                                                                                                                                                                        |                                                                                                                                                                        |
| RedpandaNodeOperationsSmokeTest                                            | test_node_ops_smoke_test                                                   | {"cloud_storage_type": 1, "mixed_versions": false} | integration | https://buildkite.com/redpanda/redpanda/builds/75218#019a30d9-5c5d-4902-bd55-798236739cdc | FLAKY         | 11/21    | upstream reliability is '100.0'. current run reliability is '52.38095238095239'. drift is 47.61905 and the allowed drift is set to 50. The test should PASS            | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test |
| RedpandaNodeOperationsSmokeTest                                            | test_node_ops_smoke_test                                                   | {"cloud_storage_type": 1, "mixed_versions": true}  | integration | https://buildkite.com/redpanda/redpanda/builds/75218#019a30d9-5c5f-4731-97bf-e30df1d19f23 | FAIL          | 0/1      |                                                                                                                                                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test |
| ScalingUpTest                                                              | test_fast_node_addition                                                    | null                                               | integration | https://buildkite.com/redpanda/redpanda/builds/75218#019a30d9-5c5c-40e6-bb43-867f2f1fef4a | FLAKY         | 19/21    | upstream reliability is '99.57627118644068'. current run reliability is '90.47619047619048'. drift is 9.10008 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ScalingUpTest&test_method=test_fast_node_addition                    |
| src/v/cloud_storage/tests/cloud_storage_e2e_test                           | src/v/cloud_storage/tests/cloud_storage_e2e_test                           |                                                    | unit        | https://buildkite.com/redpanda/redpanda/builds/75218#019a309f-5e49-458f-9734-29dc7d1c5fba | FAIL          | 0/1      |                                                                                                                                                                        |                                                                                                                                                                        |
| src/v/cloud_storage/tests/topic_recovery_service_test                      | src/v/cloud_storage/tests/topic_recovery_service_test                      |                                                    | unit        | https://buildkite.com/redpanda/redpanda/builds/75218#019a309f-5e49-458f-9734-29dc7d1c5fba | FAIL          | 0/1      |                                                                                                                                                                        |                                                                                                                                                                        |
| src/v/cloud_topics/level_one/metastore/tests/replicated_metastore_test     | src/v/cloud_topics/level_one/metastore/tests/replicated_metastore_test     |                                                    | unit        | https://buildkite.com/redpanda/redpanda/builds/75218#019a309f-5e49-458f-9734-29dc7d1c5fba | FAIL          | 0/1      |                                                                                                                                                                        |                                                                                                                                                                        |
| src/v/cluster/archival/tests/ntp_archiver_reupload_test                    | src/v/cluster/archival/tests/ntp_archiver_reupload_test                    |                                                    | unit        | https://buildkite.com/redpanda/redpanda/builds/75218#019a309f-5e49-458f-9734-29dc7d1c5fba | FAIL          | 0/1      |                                                                                                                                                                        |                                                                                                                                                                        |
| src/v/cluster/archival/tests/ntp_archiver_test                             | src/v/cluster/archival/tests/ntp_archiver_test                             |                                                    | unit        | https://buildkite.com/redpanda/redpanda/builds/75218#019a309f-5e49-458f-9734-29dc7d1c5fba | FAIL          | 0/1      |                                                                                                                                                                        |                                                                                                                                                                        |
| src/v/cluster/cloud_metadata/tests/controller_snapshot_reconciliation_test | src/v/cluster/cloud_metadata/tests/controller_snapshot_reconciliation_test |                                                    | unit        | https://buildkite.com/redpanda/redpanda/builds/75218#019a309f-5e49-458f-9734-29dc7d1c5fba | FAIL          | 0/1      |                                                                                                                                                                        |                                                                                                                                                                        |
| src/v/cluster/cloud_metadata/tests/offsets_lookup_test                     | src/v/cluster/cloud_metadata/tests/offsets_lookup_test                     |                                                    | unit        | https://buildkite.com/redpanda/redpanda/builds/75218#019a309f-5e49-458f-9734-29dc7d1c5fba | FAIL          | 0/1      |                                                                                                                                                                        |                                                                                                                                                                        |
| src/v/cluster/cloud_metadata/tests/uploader_test                           | src/v/cluster/cloud_metadata/tests/uploader_test                           |                                                    | unit        | https://buildkite.com/redpanda/redpanda/builds/75218#019a309f-5e49-458f-9734-29dc7d1c5fba | FAIL          | 0/1      |                                                                                                                                                                        |                                                                                                                                                                        |
| src/v/cluster/tests/controller_state_test                                  | src/v/cluster/tests/controller_state_test                                  |                                                    | unit        | https://buildkite.com/redpanda/redpanda/builds/75218#019a309f-5e49-458f-9734-29dc7d1c5fba | FAIL          | 0/1      |                                                                                                                                                                        |                                                                                                                                                                        |
| src/v/datalake/tests/translator_fixture_test                               | src/v/datalake/tests/translator_fixture_test                               |                                                    | unit        | https://buildkite.com/redpanda/redpanda/builds/75218#019a309f-5e49-458f-9734-29dc7d1c5fba | FAIL          | 0/1      |                                                                                                                                                                        |                                                                                                                                                                        |
| src/v/kafka/client/test/cluster_test                                       | src/v/kafka/client/test/cluster_test                                       |                                                    | unit        | https://buildkite.com/redpanda/redpanda/builds/75218#019a309f-5e49-458f-9734-29dc7d1c5fba | FAIL          | 0/1      |                                                                                                                                                                        |                                                                                                                                                                        |
| src/v/kafka/server/tests/alter_config_test                                 | src/v/kafka/server/tests/alter_config_test                                 |                                                    | unit        | https://buildkite.com/redpanda/redpanda/builds/75218#019a309f-5e49-458f-9734-29dc7d1c5fba | FAIL          | 0/1      |                                                                                                                                                                        |                                                                                                                                                                        |
| src/v/kafka/server/tests/alter_user_scram_credentials_test                 | src/v/kafka/server/tests/alter_user_scram_credentials_test                 |                                                    | unit        | https://buildkite.com/redpanda/redpanda/builds/75218#019a309f-5e49-458f-9734-29dc7d1c5fba | FAIL          | 0/1      |                                                                                                                                                                        |                                                                                                                                                                        |
| src/v/kafka/server/tests/consumer_groups_test                              | src/v/kafka/server/tests/consumer_groups_test                              |                                                    | unit        | https://buildkite.com/redpanda/redpanda/builds/75218#019a309f-5e49-458f-9734-29dc7d1c5fba | FAIL          | 0/1      |                                                                                                                                                                        |                                                                                                                                                                        |
| src/v/kafka/server/tests/kafka_fetch_plan_rpbench_test                     | src/v/kafka/server/tests/kafka_fetch_plan_rpbench_test                     |                                                    | unit        | https://buildkite.com/redpanda/redpanda/builds/75218#019a309f-5e49-458f-9734-29dc7d1c5fba | FAIL          | 0/1      |                                                                                                                                                                        |                                                                                                                                                                        |
| src/v/kafka/server/tests/write_at_offset_stm_test                          | src/v/kafka/server/tests/write_at_offset_stm_test                          |                                                    | unit        | https://buildkite.com/redpanda/redpanda/builds/75218#019a309f-5e49-458f-9734-29dc7d1c5fba | FAIL          | 0/1      |                                                                                                                                                                        |                                                                                                                                                                        |
| src/v/pandaproxy/schema_registry/test/api_rpbench_test                     | src/v/pandaproxy/schema_registry/test/api_rpbench_test                     |                                                    | unit        | https://buildkite.com/redpanda/redpanda/builds/75218#019a309f-5e49-458f-9734-29dc7d1c5fba | FAIL          | 0/1      |                                                                                                                                                                        |                                                                                                                                                                        |
| src/v/raft/tests/basic_raft_fixture_test                                   | src/v/raft/tests/basic_raft_fixture_test                                   |                                                    | unit        | https://buildkite.com/redpanda/redpanda/builds/75218#019a309f-5e49-458f-9734-29dc7d1c5fba | FAIL          | 0/1      |                                                                                                                                                                        |                                                                                                                                                                        |
| src/v/raft/tests/raft_replicate_rpbench_test                               | src/v/raft/tests/raft_replicate_rpbench_test                               |                                                    | unit        | https://buildkite.com/redpanda/redpanda/builds/75218#019a309f-5e49-458f-9734-29dc7d1c5fba | FAIL          | 0/1      |                                                                                                                                                                        |                                                                                                                                                                        |
| src/v/security/audit/tests/audit_test                                      | src/v/security/audit/tests/audit_test                                      |                                                    | unit        | https://buildkite.com/redpanda/redpanda/builds/75218#019a309f-5e49-458f-9734-29dc7d1c5fba | FAIL          | 0/1      |                                                                                                                                                                        |                                                                                                                                                                        |
| src/v/storage/tests/compaction_e2e_test                                    | src/v/storage/tests/compaction_e2e_test                                    |                                                    | unit        | https://buildkite.com/redpanda/redpanda/builds/75218#019a309f-5e49-458f-9734-29dc7d1c5fba | FAIL          | 0/1      |                                                                                                                                                                        |                                                                                                                                                                        |
| src/v/storage/tests/segment_appender_rpbench_test                          | src/v/storage/tests/segment_appender_rpbench_test                          |                                                    | unit        | https://buildkite.com/redpanda/redpanda/builds/75218#019a309f-5e49-458f-9734-29dc7d1c5fba | FAIL          | 0/1      |                                                                                                                                                                        |                                                                                                                                                                        |
| src/v/storage/tests/segment_appender_test                                  | src/v/storage/tests/segment_appender_test                                  |                                                    | unit        | https://buildkite.com/redpanda/redpanda/builds/75218#019a309f-5e49-458f-9734-29dc7d1c5fba | FAIL          | 0/1      |                                                                                                                                                                        |                                                                                                                                                                        |
| src/v/storage/tests/storage_e2e_fixture_test                               | src/v/storage/tests/storage_e2e_fixture_test                               |                                                    | unit        | https://buildkite.com/redpanda/redpanda/builds/75218#019a309f-5e49-458f-9734-29dc7d1c5fba | FAIL          | 0/1      |                                                                                                                                                                        |                                                                                                                                                                        |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75299">build#75299</a></summary>


| test_class                                        | test_method                                       | test_arguments              | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                     |
|:--------------------------------------------------|:--------------------------------------------------|:----------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTxTest                         | test_write                                        | null                        | integration | https://buildkite.com/redpanda/redpanda/builds/75299#019a3471-576c-442c-9853-6c2cae3180b3 | FLAKY         | 17/21    | upstream reliability is '93.40974212034384'. current run reliability is '80.95238095238095'. drift is 12.45736 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=EndToEndCloudTopicsTxTest&test_method=test_write               |
| WriteCachingFailureInjectionE2ETest               | test_crash_all                                    | {"use_transactions": false} | integration | https://buildkite.com/redpanda/redpanda/builds/75299#019a3471-576f-4113-98c4-af69903f811d | FLAKY         | 19/21    | upstream reliability is '84.4017094017094'. current run reliability is '90.47619047619048'. drift is -6.07448 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all |
| src/v/storage/tests/segment_appender_rpbench_test | src/v/storage/tests/segment_appender_rpbench_test |                             | unit        | https://buildkite.com/redpanda/redpanda/builds/75299#019a3438-9d32-47e4-9cfe-dbfbe91e6a02 | FAIL          | 0/1      |                                                                                                                                                                         |                                                                                                                                                                  |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75472">build#75472</a></summary>


| test_class                    | test_method                       | test_arguments                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                                  |
|:------------------------------|:----------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| AuditLogTestKafkaApi          | test_no_auth_enabled              | {"audit_transport_mode": "rpc"}                                                                                                                | integration | https://buildkite.com/redpanda/redpanda/builds/75472#019a499c-2616-4c84-b026-3240432cad91 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=AuditLogTestKafkaApi&test_method=test_no_auth_enabled                       |
| ShadowLinkingReplicationTests | test_replication_basic            | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "redpanda"}}                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/75472#019a499c-2609-41a5-b4f3-006930ca3852 | FLAKY         | 20/21    | upstream reliability is '97.9689366786141'. current run reliability is '95.23809523809523'. drift is 2.73084 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic            |
| ShadowLinkingReplicationTests | test_replication_timestamps_match | {"source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}, "timestamp_type": "CreateTime"} | integration | https://buildkite.com/redpanda/redpanda/builds/75472#019a4995-ad94-441c-a99f-844c11a03f4e | FLAKY         | 20/21    | upstream reliability is '95.54794520547945'. current run reliability is '95.23809523809523'. drift is 0.30985 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_timestamps_match |
| ShadowLinkingReplicationTests | test_replication_timestamps_match | {"source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}, "timestamp_type": "CreateTime"} | integration | https://buildkite.com/redpanda/redpanda/builds/75472#019a499c-260a-4486-a755-c63906aa85ab | FLAKY         | 20/21    | upstream reliability is '95.54794520547945'. current run reliability is '95.23809523809523'. drift is 0.30985 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_timestamps_match |
| QuickTerminateTest            | test_terminate                    | null                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/75472#019a499c-2619-463a-bf8b-1af290c65052 | FAIL          | 0/1      |                                                                                                                                                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=QuickTerminateTest&test_method=test_terminate                               |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75551">build#75551</a></summary>


| test_class                          | test_method                   | test_arguments              | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                           |
|:------------------------------------|:------------------------------|:----------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTxTest           | test_write                    | null                        | integration | https://buildkite.com/redpanda/redpanda/builds/75551#019a4f9b-54a3-441c-9862-d3032ec4e607 | FLAKY         | 16/21    | upstream reliability is '91.05691056910568'. current run reliability is '76.19047619047619'. drift is 14.86643 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=EndToEndCloudTopicsTxTest&test_method=test_write                     |
| LogCompactionTxRemovalTest          | test_tx_control_batch_removal | null                        | integration | https://buildkite.com/redpanda/redpanda/builds/75551#019a4f9b-54a4-4683-b5b7-9de6da4dc901 | FLAKY         | 14/21    | upstream reliability is '87.25868725868726'. current run reliability is '66.66666666666666'. drift is 20.59202 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=LogCompactionTxRemovalTest&test_method=test_tx_control_batch_removal |
| WriteCachingFailureInjectionE2ETest | test_crash_all                | {"use_transactions": false} | integration | https://buildkite.com/redpanda/redpanda/builds/75551#019a4fa1-0302-4403-8979-a1561180a412 | FLAKY         | 19/21    | upstream reliability is '87.70144189991518'. current run reliability is '90.47619047619048'. drift is -2.77475 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all       |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77914">build#77914</a></summary>


| test_class             | test_method                                | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                                 |
|:-----------------------|:-------------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| SegmentAppenderFixture | TestFlushesAreMerged                       |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/77914#019b24a1-417b-43ef-807a-cd4ae1d877b4 | FAIL          | 0/1      |                                                                                                                                                                                                                                                                                                                                                                                                     |                                                                                                                                                                              |
| CrashLoopChecksTest    | test_cloud_storage_client_misconfiguration | null             | integration | https://buildkite.com/redpanda/redpanda/builds/77914#019b24d5-dda6-49e9-82eb-1ed57ed181c3 | FLAKY         | 3/11     | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.9265, p0=0.9957, reject_threshold=0.0100. adj_baseline=0.9996, p1=0.0000, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=CrashLoopChecksTest&test_method=test_cloud_storage_client_misconfiguration |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/78528">build#78528</a></summary>


| test_class                          | test_method                          | test_arguments                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                                  |
|:------------------------------------|:-------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests       | test_replication_timestamps_match    | {"source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}, "timestamp_type": "CreateTime"} | integration | https://buildkite.com/redpanda/redpanda/builds/78528#019b8d4d-7143-413b-9e70-f4b5f5f3eb20 | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0000, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.1000, p1=0.3487, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_timestamps_match |
| DatalakeBlockedCatalogTest          | test_disable_reenable_after_blocking | {"cloud_storage_type": 1}                                                                                                                      | integration | https://buildkite.com/redpanda/redpanda/builds/78528#019b8d4d-7143-413b-9e70-f4b5f5f3eb20 | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0000, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.1000, p1=0.3487, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeBlockedCatalogTest&test_method=test_disable_reenable_after_blocking |
| WriteCachingFailureInjectionE2ETest | test_crash_all                       | {"use_transactions": false}                                                                                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/78528#019b8d4d-0b5f-433a-ac51-a6524a749ac0 | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0931, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.2542, p1=0.0532, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all              |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/79256">build#79256</a></summary>


| test_class                          | test_method              | test_arguments                                     | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                           |
|:------------------------------------|:-------------------------|:---------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingMetricsTests           | test_link_metrics        | null                                               | integration | https://buildkite.com/redpanda/redpanda/builds/79256#019bd6e6-5935-4bfa-abeb-97806bc33f3b | FLAKY         | 9/11     | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.1096, p0=0.6867, reject_threshold=0.0100. adj_baseline=0.2940, p1=0.1589, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingMetricsTests&test_method=test_link_metrics              |
| MountUnmountIcebergTest             | test_simple_remount      | {"cloud_storage_type": 1}                          | integration | https://buildkite.com/redpanda/redpanda/builds/79256#019bd6ed-2b5a-4336-915f-c15fd77831ec | FLAKY         | 7/11     | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.1692, p0=0.2320, reject_threshold=0.0100. adj_baseline=0.4266, p1=0.3179, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MountUnmountIcebergTest&test_method=test_simple_remount              |
| RedpandaNodeOperationsSmokeTest     | test_node_ops_smoke_test | {"cloud_storage_type": 1, "mixed_versions": false} | integration | https://buildkite.com/redpanda/redpanda/builds/79256#019bd6e6-593d-46d7-9171-7a2c79b6e8fe | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0748, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.2081, p1=0.0970, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test |
| WriteCachingFailureInjectionE2ETest | test_crash_all           | {"use_transactions": false}                        | integration | https://buildkite.com/redpanda/redpanda/builds/79256#019bd6ed-2b5d-41de-861a-f5e055308f97 | FLAKY         | 23/31    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0981, p0=0.0235, reject_threshold=0.0100. adj_baseline=0.2664, p1=0.4321, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all       |
</details>
<!--/body-->

mmaslankaprv commented:
/dt

mmaslankaprv commented:
/dt

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/75472-->
<!--218cd32b-38f6-4e80-9c6c-bdfa8a3fe465-->
<!--title-->
## Retry command for [Build#75472](https://buildkite.com/redpanda/redpanda/builds/75472)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/quick_terminate_test.py::QuickTerminateTest.test_terminate
```
<!--/body-->

mmaslankaprv commented:
/ci-repeat 1

dotnwat commented on src/v/storage/segment_appender_chunk.h:
```diff
@@ -105,6 +105,25 @@ class segment_appender_chunk {
+     */
+    void copy_reminder_from(const segment_appender_chunk& other) {
+        reset();
+        const auto reminder_sz = other.size() - other.pending_aligned_begin();
```
nit: i think this is intended to be "remainder" ?

dotnwat commented on src/v/storage/segment_appender.cc:
```diff
@@ -165,6 +165,61 @@ ss::future<> segment_appender::do_append(const char* buf, size_t n) {
         }
+        // we need to copy the reminder of the chunk into the new one not write
+        // to the one currently being written
+        if (is_chunk_write_dispatched(_head)) {
```
i kind of imagined that we'd "split" _head (logically). there would be something like _head buffer as it exists today with shared ownership among append/truncate/flush etc... then the write dispatching would form the other half and operate on a exclusive ownership buffer.

append/truncate/flush/etc... would move a copy of their buffer into the write dispatch path (if the buffer were not full) or move the whole thing (if the write filled up the buffer).

i suspect from looking that this is similar, but it seems like we are adding complexity in order to keep _head in place as a shared ownership member across all these code paths?

dotnwat commented:
/cdt
rp_version=build

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/78474-->
<!--db8fa813-f366-4051-b7af-0a2d3e86cfee-->
<!--60e0956a-5675-48bd-8ee7-10b687879dd2-->
<!--title-->
## Retry command for [Build#78474](https://buildkite.com/redpanda/redpanda/builds/78474)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
skip-redpanda-build
skip-units
skip-rebase
tests/rptest/tests/datalake/datalake_dlq_test.py::DatalakeDLQTest.test_dlq_table_for_mixed_records@{"catalog_type":"rest_jdbc","cloud_storage_type":1,"query_engine":"spark"}
tests/rptest/tests/nodes_decommissioning_test.py::NodesDecommissioningTest.test_multiple_decommissions@{"cloud_topic":true}
```
<!--/body-->

dotnwat commented:
/cdt

mmaslankaprv committed:
s/tests: removed overly strict validation in appender test

Current segment appender implementation may sometimes execute more than
expected number of fsync operations. This was the case before. Relaxed
the test expectations to make the test stable.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

vbotbuildovich commented:
<!--0c8afd36-d8f1-40e7-b53e-af7b286fd53b-->
<!--0b19ddd2-cbba-4102-8e11-273b8c142dbd-->
<!--body-->
Performance change detected in https://buildkite.com/redpanda/redpanda/builds/78586#019b93ec-eafb-4067-874c-587a729ae2f8:
```
Performance changes detected in 9 tests
cloud_storage_rpbench.cstore_bench.cs_iteration_recompute_end_test_1000: inst -> +0.14pct
cloud_storage_rpbench.cstore_bench.cs_iteration_recompute_end_test_10000: inst -> +0.15pct
segment_appender_rpbench.appender_fixture.128_KiB_writes_flush: inst -> +0.06pct
segment_appender_rpbench.appender_fixture.128_KiB_writes_no_flush: inst -> +0.06pct
segment_appender_rpbench.appender_fixture.1_byte_writes_flush: inst -> +0.26pct
segment_appender_rpbench.appender_fixture.1_byte_writes_no_flush: inst -> +0.31pct
segment_appender_rpbench.appender_fixture.4096_byte_writes_flush: inst -> +0.11pct
segment_appender_rpbench.appender_fixture.4096_byte_writes_no_flush: inst -> +0.06pct
segment_appender_rpbench.appender_fixture.97_byte_writes_flush: inst -> +0.26pct
```
See https://redpandadata.atlassian.net/wiki/x/LQAqLg for docs
<!--/body-->

mmaslankaprv commented on src/v/storage/segment_appender.cc:
```diff
@@ -165,6 +165,61 @@ ss::future<> segment_appender::do_append(const char* buf, size_t n) {
+
+                auto new_head = co_await internal::chunks().get();
+
+                const auto reminder_sz = old_head->size()
```
on line 25 there is an `exchange` call. So the `_head` is not 'attached' to old head before the scheduling point therefore we can assume the `old_head` will not change
